add support for useImplementingTypes

maordaniel opened the pull request.

ardeois commented on src/index.ts:
```diff
@@ -544,6 +575,32 @@ export const plugin: PluginFunction<TypescriptMocksPluginConfig> = (schema, docu
+        },
+        ScalarTypeDefinition: (node) => {
+            const name = node.name.value;
+            if (!types.find((enumType) => enumType.name === name)) {
```
rename `enumType` to `scalarType` ?

ardeois commented:
Thanks for implementing this @maordaniel 

I've got a few comments and the unit tests are failing

maordaniel commented on src/index.ts:
```diff
@@ -300,23 +325,27 @@ const getNamedType = (opts: Options<NamedTypeNode>): string | number | boolean =
-                        )}({}, relationshipsToOmit)`,
-                );
+                return `relationshipsToOmit.has('${casedName}') ? {} as ${casedName} : ${toMockName(
+                    name,
+                    casedName,
+                    opts.prefix,
+                )}({}, relationshipsToOmit)`;
             } else {
-                return handleValueGeneration(opts, null, () => `${toMockName(name, casedName, opts.prefix)}()`);
+                return `${toMockName(name, casedName, opts.prefix)}()`;
```
no, sorry about that.
I will fix it.
do i need to use "handleValueGeneration" in getNamedImplementType func?

ardeois replied:
Yes probably, `handleValueGeneration` is necessary when you use config `fieldGeneration` 
Here is a unit test example. You should add one to make sure it works with implementing types
https://github.com/ardeois/graphql-codegen-typescript-mock-data/blob/5b1038787da21f80726a5f5d1a79408e0cd5a91a/tests/perTypeFieldGeneration/spec.ts#L46-L65

maordaniel replied:
look ... I don't think it is related, nor is it necessary.
if someone using fieldGeneration, why he use useImplementingTypes also?
This is contrary to the idea.

maordaniel replied:
for example:

```
const result = await plugin(testSchema, [], { 
         ...config, 
         fieldGeneration: { 
             A: { config: 'internet.config' }, 
         }, 
     }); 
    
 expect(result).toContain(
        "config: overrides && overrides.hasOwnProperty('config') ? overrides.config! : internet.config() || internet.config(),",
    );
```

maordaniel replied:
Ok, I've fix it.
Now if the prop useimplementingtypes is true and the fieldgeneration is not empty, the code will ignore the implant types.. i added test about that

maordaniel commented:
> Thanks for implementing this @maordaniel
> 
> I've got a few comments and the unit tests are failing

 The unit tests that fail are not related to changes in my pr.. @ardeois

maordaniel commented:
> @maordaniel tests related to your code are now failing ![CleanShot 2023-03-23 at 15 36 27](https://user-images.githubusercontent.com/1867939/227331378-39c5fc52-9711-4b37-b59c-3cd6e7c73e98.png)

fixed.

maordaniel commented:
> @maordaniel tests related to your code are now failing
> ![CleanShot 2023-03-23 at 15 36 27](https://user-images.githubusercontent.com/1867939/227331378-39c5fc52-9711-4b37-b59c-3cd6e7c73e98.png)
> 
> 
> 
> 

Fixed and i added one more test

ardeois commented on tests/useImplementingTypes/__snapshots__/spec.ts.snap:
```diff
@@ -0,0 +1,247 @@
+"
+`;
+
+exports[`shouldn't support useImplementingTypes 2`] = `
```
Seems like this snapshot is here 2 times. Are you sure you don't have obsolete snapshots?

maordaniel replied:
fixed

ardeois commented on tests/useImplementingTypes/__snapshots__/spec.ts.snap:
```diff
@@ -0,0 +1,247 @@
+"
+`;
+
+exports[`shouldn't support useImplementingTypes with fieldGeneration props 1`] = `
```
Same here the snapshot is here twice 🤔

ardeois commented on tests/useImplementingTypes/spec.ts:
```diff
@@ -0,0 +1,43 @@
+it(`shouldn't support useImplementingTypes with fieldGeneration prop`, async () => {
+    const result = await plugin(testSchema, [], {
+        prefix: 'mock',
+        useImplementingTypes: true,
+        fieldGeneration: {
+            A: { config: 'email' },
+        },
```
What if we want to use `useImplementingTypes` (for `config`) and want to use `fieldGeneration` for another field.
Like:
```
        useImplementingTypes: true,
        fieldGeneration: {
            A: { str: 'email' },
        },
```


I think that could make sense, but your code will disable implementing types if `fieldGeneration` is defined.
You should instead disable `useImplementingTypes` only for properties that are overridden by `fieldGeneration

maordaniel replied:
done.

ardeois reviewed (commented):
Sorry for the late review, I got quite busy these last few days.
I tried to do an in depth review and found potential issues.
Let me know what you think

Thanks again for the help !

maord-upstream commented on tests/useImplementingTypes/__snapshots__/spec.ts.snap:
```diff
@@ -0,0 +1,247 @@
+};
+
+export const mockTestAConfig = (overrides?: Partial<TestAConfig>): TestAConfig => {
+    return {
+        detectionTypes: overrides && overrides.hasOwnProperty('detectionTypes') ? overrides.detectionTypes! : [ConfigTypes.Test],
+        active: overrides && overrides.hasOwnProperty('active') ? overrides.active! : true,
+    };
+};
```
yes, i just need to fix my schema...

maordaniel replied:
yes you right , i've fixed the schema.

maordaniel commented:
> Sorry for the late review, I got quite busy these last few days. I tried to do an in depth review and found potential issues. Let me know what you think
> 
> Thanks again for the help !

Hi Ardeois,
hope you are well.
i've fixed the issues you mentioned.
thanks 🙏

ardeois reviewed (approved):
Looks good ! thanks for the updates

maordaniel commented:
> Looks good ! thanks for the updates

Thanks a lot. 
Really appreciate that ! @ardeois 

I updated the snapshot for the test, pls approve this again.

maordaniel commented:
Hi @ardeois ,
When are you going to merge it?

Thanks

maordaniel commented:
Thanks a lot! 🙏🏻 @ardeois
